---
title: "1. Presentation"
---

_Jacques Serizay_  

## Dates

Jan./Feb. 2022

## Where

This course will be held online.

## Overview

This course will introduce biologists and bioinformaticians to the 
field of single-cell RNA sequencing. 
We will cover a range of software and analysis workflows that extend 
over the spectrum from the best practices in the filtering scRNA-seq data  
to the downstream analysis of cell clusters and temporal ordering. 
This course will help the attendees gain accurate insights in pre-processing, 
analysis and interpretation of scRNA-seq data.

We will start by introducing general concepts about single-cell RNA-sequencing. 
From there, we will then continue to describe the main analysis steps to go from 
raw sequencing data to processed and usable data. We will present classical 
analysis workflows, their output and the possible paths to investigate for 
downstream analysis.  

Throughout this workshop, we will put an emphasis on `R/Bioconductor` ecosystem 
and the different packages which will be used to analyse datasets and learn new approaches. 

## Format

The course is structured in modules over five days.  

During the first 1/3 of the day, formal lectures will cover the key 
concepts required to understand the principles of scRNA-seq analysis (~2h).

Following these lectures, practical examples will be shown to illustrate how 
to translate the acquired knowledge into functional `R` code (~1h). At this stage, 
trainees will get acquainted with state-of-the-art packages for scRNAseq analysis 
as well as the best practices in bioinformatics. 

During the second half of the day (3h), trainees will work by themselves, following 
guided exercises to improve their understanding of scRNAseq analysis workflow.
Hints and solution are provided for each exercise. The exercises will mainly focus 
on specific concepts introducted earlier that day.
However, analytical steps studied throughout the previous days will also 
be integrated so that towards the end of the week, trainees are fully able 
to perform fundamental scRNAseq analyses from beginning to end. 

Office hours will take place during the last hour of the exercises. An 
instructor will be available to answer individual questions related to daily exercises. 
A Slack channel will also be available so that Q&A are available for everybody.

<!-- ## Targeted audience and required background

The course will be mostly beneficial to those who have, or will shortly have, scRNA-seq data ready to analyse.  

The material is suitable both for experimentalists who want to learn 
more about data-analysis as well as computational biologists 
who want to learn about scRNA-seq methods.  

Examples demonstrated in this course can be applied to any experimental protocol or biological system.

The requirements for this course are: 

1. Working knowledge of Unix / command line interface (managing files, running programs, reading manuals!). Basic bash commands (`cd`, `ls`, `...`) and CLI usage will _not_ be covered in this course. We advice attendees to not register if they lack fundamental experience in CLI.
2. Programming experience in `R` (writing a function, basic I/O operations, variable types, using packages). `Bioconductor` experience is a plus. 
3. Familiarity with next-generation sequencing data and its analyses (using alignment and quantification tools for bulk sequencing data) -->

## Learning outcomes

At the end of this course, you should be able to:

- Understand the pros/cons of different single-cell RNA-seq methods
- Process and QC of scRNA-seq data
- Normalize scRNA-seq data
- Correct for batch effects
- Visualise the data and applying dimensionality reduction
- Perform cell clustering and annotation 
- Perform differential gene expression analysis
- Infer cell trajectory and pseudotime, and perform temporal differential expression 

--- 

## Instructors

Dr. Jacques Serizay  
